Add item requests

Vexvl opened the pull request.

Vexvl committed:
feat: ItemRequestController.java
      ItemRequestDto.java
      ItemRequestMapper.java
      ItemRequestRepository.java
      ItemRequestService.java
      ItemRequestServiceImpl.java

fix and feat: Booking.java
              BookingController.java
              BookingMapper.java
              BookingRepository.java
              BookingService.java
              BookingServiceImpl.java
              CommentRepository.java
              CommentServiceImpl.java
              ItemController.java
              ItemDto.java
              ItemMapper.java
              ItemRepository.java
              ItemRequest.java
              ItemService.java
              ItemServiceImpl.java
              schema.sql

Vexvl committed:
feat: BookingControllerTests.java
      BookingMapperTests.java
      BookingRepositoryTest.java
      BookingServiceImpl.java
      BookingTest.java
      BookingTests.java
      CommentDto.java
      CommentMapperTests.java
      CommentTests.java
      ExceptionsHandler.java
      InvalidStatusException.java
      ItemControllerTest.java
      ItemMapperTests.java
      ItemRepositoryTest.java
      ItemRequestControllerTests.java
      ItemRequestMapperTests.java
      ItemTest.java
      ItemTests.java
      pom.xml
      RequestRepositoryTest.java
      RequestTest.java
      UnsupportedStatusException.java
      UserControllerTests.java
      UserDto.java
      UserMapperTests.java
      UserTest.java

Vexvl committed:
feat: BookingControllerTests.java
      BookingMapperTests.java
      BookingRepositoryTest.java
      BookingServiceImpl.java
      BookingTest.java
      BookingTests.java
      CommentDto.java
      CommentMapperTests.java
      CommentTests.java
      ExceptionsHandler.java
      InvalidStatusException.java
      ItemControllerTest.java
      ItemMapperTests.java
      ItemRepositoryTest.java
      ItemRequestControllerTests.java
      ItemRequestMapperTests.java
      ItemTest.java
      ItemTests.java
      pom.xml
      RequestRepositoryTest.java
      RequestTest.java
      UnsupportedStatusException.java
      UserControllerTests.java
      UserDto.java
      UserMapperTests.java
      UserTest.java

Vexvl committed:
feat: BookingControllerTests.java
      BookingMapperTests.java
      BookingRepositoryTest.java
      BookingServiceImpl.java
      BookingTest.java
      BookingTests.java
      CommentDto.java
      CommentMapperTests.java
      CommentTests.java
      ExceptionsHandler.java
      InvalidStatusException.java
      ItemControllerTest.java
      ItemMapperTests.java
      ItemRepositoryTest.java
      ItemRequestControllerTests.java
      ItemRequestMapperTests.java
      ItemTest.java
      ItemTests.java
      pom.xml
      RequestRepositoryTest.java
      RequestTest.java
      UnsupportedStatusException.java
      UserControllerTests.java
      UserDto.java
      UserMapperTests.java
      UserTest.java

Vexvl committed:
feat: BookingControllerTests.java
      BookingMapperTests.java
      BookingRepositoryTest.java
      BookingServiceImpl.java
      BookingTest.java
      BookingTests.java
      CommentDto.java
      CommentMapperTests.java
      CommentTests.java
      ExceptionsHandler.java
      InvalidStatusException.java
      ItemControllerTest.java
      ItemMapperTests.java
      ItemRepositoryTest.java
      ItemRequestControllerTests.java
      ItemRequestMapperTests.java
      ItemTest.java
      ItemTests.java
      pom.xml
      RequestRepositoryTest.java
      RequestTest.java
      UnsupportedStatusException.java
      UserControllerTests.java
      UserDto.java
      UserMapperTests.java
      UserTest.java

Vexvl committed:
feat: BookingControllerTests.java
      BookingMapperTests.java
      BookingRepositoryTest.java
      BookingServiceImpl.java
      BookingTest.java
      BookingTests.java
      CommentDto.java
      CommentMapperTests.java
      CommentTests.java
      ExceptionsHandler.java
      InvalidStatusException.java
      ItemControllerTest.java
      ItemMapperTests.java
      ItemRepositoryTest.java
      ItemRequestControllerTests.java
      ItemRequestMapperTests.java
      ItemTest.java
      ItemTests.java
      pom.xml
      RequestRepositoryTest.java
      RequestTest.java
      UnsupportedStatusException.java
      UserControllerTests.java
      UserDto.java
      UserMapperTests.java
      UserTest.java

Vexvl committed:
feat: BookingControllerTests.java
      BookingMapperTests.java
      BookingRepositoryTest.java
      BookingServiceImpl.java
      BookingTest.java
      BookingTests.java
      CommentDto.java
      CommentMapperTests.java
      CommentTests.java
      ExceptionsHandler.java
      InvalidStatusException.java
      ItemControllerTest.java
      ItemMapperTests.java
      ItemRepositoryTest.java
      ItemRequestControllerTests.java
      ItemRequestMapperTests.java
      ItemTest.java
      ItemTests.java
      pom.xml
      RequestRepositoryTest.java
      RequestTest.java
      UnsupportedStatusException.java
      UserControllerTests.java
      UserDto.java
      UserMapperTests.java
      UserTest.java

Vexvl committed:
feat: BookingControllerTests.java
      BookingMapperTests.java
      BookingRepositoryTest.java
      BookingServiceImpl.java
      BookingTest.java
      BookingTests.java
      CommentDto.java
      CommentMapperTests.java
      CommentTests.java
      ExceptionsHandler.java
      InvalidStatusException.java
      ItemControllerTest.java
      ItemMapperTests.java
      ItemRepositoryTest.java
      ItemRequestControllerTests.java
      ItemRequestMapperTests.java
      ItemTest.java
      ItemTests.java
      pom.xml
      RequestRepositoryTest.java
      RequestTest.java
      UnsupportedStatusException.java
      UserControllerTests.java
      UserDto.java
      UserMapperTests.java
      UserTest.java

Vexvl committed:
feat: BookingRepository.java
      BookingRepositoryTest.java
      BookingTest.java
      BookingTests.java
      ItemRequestControllerTests.java
      ItemTests.java
      UserControllerTests.java

Vexvl committed:
feat: BookingRepository.java
      BookingRepositoryTest.java
      BookingTest.java
      BookingTests.java
      ItemRequestControllerTests.java
      ItemTests.java
      UserControllerTests.java

Vexvl committed:
feat:BookingControllerTests.java
     BookingIntegrationTests.java
     BookingMapperTests.java
     BookingRepositoryTest.java
     BookingUnitTests.java
     CommentControllerTests.java
     CommentIntegrationTests.java
     CommentMapperTests.java
     CommentRepositoryTests.java
     CommentUnitTests.java
     ItemControllerTest.java
     ItemIntegrationTests.java
     ItemMapperTests.java
     ItemRepositoryTest.java
     ItemRequestControllerTests.java
     ItemRequestMapperTests.java
     ItemRequestServiceImpl.java
     ItemRequestTest.java
     ItemTest.java
     ItemUnitTests.java
     RequestRepositoryTest.java
     UserControllerTests.java
     UserIntegrationTests.java
     UserMapperTests.java
     UserRepositoryTests.java
     UserUnitTests.java

Vexvl committed:
feat:BookingControllerTests.java
     BookingIntegrationTests.java
     BookingMapperTests.java
     BookingRepositoryTest.java
     BookingUnitTests.java
     CommentControllerTests.java
     CommentIntegrationTests.java
     CommentMapperTests.java
     CommentRepositoryTests.java
     CommentUnitTests.java
     ItemControllerTest.java
     ItemIntegrationTests.java
     ItemMapperTests.java
     ItemRepositoryTest.java
     ItemRequestControllerTests.java
     ItemRequestMapperTests.java
     ItemRequestServiceImpl.java
     ItemRequestTest.java
     ItemTest.java
     ItemUnitTests.java
     RequestRepositoryTest.java
     UserControllerTests.java
     UserIntegrationTests.java
     UserMapperTests.java
     UserRepositoryTests.java
     UserUnitTests.java

Vexvl committed:
Merge remote-tracking branch 'origin/add-item-requests' into add-item-requests

Vexvl committed:
feat:BookingControllerTests.java
     BookingIntegrationTests.java
     BookingMapperTests.java
     BookingRepositoryTest.java
     BookingUnitTests.java
     CommentControllerTests.java
     CommentIntegrationTests.java
     CommentMapperTests.java
     CommentRepositoryTests.java
     CommentUnitTests.java
     ItemControllerTest.java
     ItemIntegrationTests.java
     ItemMapperTests.java
     ItemRepositoryTest.java
     ItemRequestControllerTests.java
     ItemRequestMapperTests.java
     ItemRequestServiceImpl.java
     ItemRequestTest.java
     ItemTest.java
     ItemUnitTests.java
     RequestRepositoryTest.java
     UserControllerTests.java
     UserIntegrationTests.java
     UserMapperTests.java
     UserRepositoryTests.java
     UserUnitTests.java

Vexvl committed:
feat:BookingControllerTests.java
     BookingIntegrationTests.java
     BookingMapperTests.java
     BookingRepositoryTest.java
     BookingUnitTests.java
     CommentControllerTests.java
     CommentIntegrationTests.java
     CommentMapperTests.java
     CommentRepositoryTests.java
     CommentUnitTests.java
     ItemControllerTest.java
     ItemIntegrationTests.java
     ItemMapperTests.java
     ItemRepositoryTest.java
     ItemRequestControllerTests.java
     ItemRequestMapperTests.java
     ItemRequestServiceImpl.java
     ItemRequestTest.java
     ItemTest.java
     ItemUnitTests.java
     RequestRepositoryTest.java
     UserControllerTests.java
     UserIntegrationTests.java
     UserMapperTests.java
     UserRepositoryTests.java
     UserUnitTests.java

Vexvl committed:
feat:BookingControllerTests.java
     BookingIntegrationTests.java
     BookingMapperTests.java
     BookingRepositoryTest.java
     BookingUnitTests.java
     CommentControllerTests.java
     CommentIntegrationTests.java
     CommentMapperTests.java
     CommentRepositoryTests.java
     CommentUnitTests.java
     ItemControllerTest.java
     ItemIntegrationTests.java
     ItemMapperTests.java
     ItemRepositoryTest.java
     ItemRequestControllerTests.java
     ItemRequestMapperTests.java
     ItemRequestServiceImpl.java
     ItemRequestTest.java
     ItemTest.java
     ItemUnitTests.java
     RequestRepositoryTest.java
     UserControllerTests.java
     UserIntegrationTests.java
     UserMapperTests.java
     UserRepositoryTests.java
     UserUnitTests.java

Vexvl committed:
feat:BookingControllerTests.java
     BookingIntegrationTests.java
     BookingMapperTests.java
     BookingRepositoryTest.java
     BookingUnitTests.java
     CommentControllerTests.java
     CommentIntegrationTests.java
     CommentMapperTests.java
     CommentRepositoryTests.java
     CommentUnitTests.java
     ItemControllerTest.java
     ItemIntegrationTests.java
     ItemMapperTests.java
     ItemRepositoryTest.java
     ItemRequestControllerTests.java
     ItemRequestMapperTests.java
     ItemRequestServiceImpl.java
     ItemRequestTest.java
     ItemTest.java
     ItemUnitTests.java
     RequestRepositoryTest.java
     UserControllerTests.java
     UserIntegrationTests.java
     UserMapperTests.java
     UserRepositoryTests.java
     UserUnitTests.java

Vexvl committed:
feat:BookingControllerTests.java
     BookingIntegrationTests.java
     BookingMapperTests.java
     BookingRepositoryTest.java
     BookingUnitTests.java
     CommentControllerTests.java
     CommentIntegrationTests.java
     CommentMapperTests.java
     CommentRepositoryTests.java
     CommentUnitTests.java
     ItemControllerTest.java
     ItemIntegrationTests.java
     ItemMapperTests.java
     ItemRepositoryTest.java
     ItemRequestControllerTests.java
     ItemRequestMapperTests.java
     ItemRequestServiceImpl.java
     ItemRequestTest.java
     ItemTest.java
     ItemUnitTests.java
     RequestRepositoryTest.java
     UserControllerTests.java
     UserIntegrationTests.java
     UserMapperTests.java
     UserRepositoryTests.java
     UserUnitTests.java

Vexvl committed:
feat:BookingControllerTests.java
     BookingIntegrationTests.java
     BookingMapperTests.java
     BookingRepositoryTest.java
     BookingUnitTests.java
     CommentControllerTests.java
     CommentIntegrationTests.java
     CommentMapperTests.java
     CommentRepositoryTests.java
     CommentUnitTests.java
     ItemControllerTest.java
     ItemIntegrationTests.java
     ItemMapperTests.java
     ItemRepositoryTest.java
     ItemRequestControllerTests.java
     ItemRequestMapperTests.java
     ItemRequestServiceImpl.java
     ItemRequestTest.java
     ItemTest.java
     ItemUnitTests.java
     RequestRepositoryTest.java
     UserControllerTests.java
     UserIntegrationTests.java
     UserMapperTests.java
     UserRepositoryTests.java
     UserUnitTests.java

Vexvl committed:
Merge remote-tracking branch 'origin/add-item-requests' into add-item-requests

Vexvl committed:
Merge remote-tracking branch 'origin/add-item-requests' into add-item-requests

Vexvl committed:
Merge remote-tracking branch 'origin/add-item-requests' into add-item-requests

Vexvl committed:
Merge remote-tracking branch 'origin/add-item-requests' into add-item-requests

Vexvl committed:
BookingIntegrationTests.java
BookingRepositoryTest.java
BookingUnitTests.java
ItemRequestUnitTest.java
ItemUnitTests.java

Vexvl committed:
Merge remote-tracking branch 'origin/add-item-requests' into add-item-requests

Vexvl committed:
BookingIntegrationTests.java
BookingRepositoryTest.java
BookingUnitTests.java
ItemRequestUnitTest.java
ItemUnitTests.java

Vexvl committed:
BookingIntegrationTests.java
BookingRepositoryTest.java
BookingUnitTests.java
ItemRequestUnitTest.java
ItemUnitTests.java

Vexvl committed:
BookingIntegrationTests.java
BookingRepositoryTest.java
BookingUnitTests.java
ItemRequestUnitTest.java
ItemUnitTests.java

Vexvl committed:
BookingIntegrationTests.java
BookingRepositoryTest.java
BookingUnitTests.java
ItemRequestUnitTest.java
ItemUnitTests.java

Vexvl committed:
BookingIntegrationTests.java
BookingRepositoryTest.java
BookingUnitTests.java
ItemRequestUnitTest.java
ItemUnitTests.java

Vexvl committed:
BookingIntegrationTests.java
BookingRepositoryTest.java
BookingUnitTests.java
ItemRequestUnitTest.java
ItemUnitTests.java

Vexvl committed:
BookingIntegrationTests.java
BookingRepositoryTest.java
BookingUnitTests.java
ItemRequestUnitTest.java
ItemUnitTests.java

Vexvl committed:
BookingIntegrationTests.java
BookingRepositoryTest.java
BookingUnitTests.java
ItemRequestUnitTest.java
ItemUnitTests.java

Vexvl committed:
BookingIntegrationTests.java
BookingRepositoryTest.java
BookingUnitTests.java
ItemRequestUnitTest.java
ItemUnitTests.java

Vexvl committed:
BookingIntegrationTests.java
BookingRepositoryTest.java
BookingUnitTests.java
ItemRequestUnitTest.java
ItemUnitTests.java

sergeymamonov commented on src/main/java/ru/practicum/shareit/comment/service/Impl/CommentServiceImpl.java:
```diff
@@ -1,4 +1,4 @@
-package ru.practicum.shareit.comment.service;
+package ru.practicum.shareit.comment.service.Impl;
```
Impl неверное наименование пакета.

sergeymamonov commented on src/main/java/ru/practicum/shareit/handler/ExceptionsHandler.java:
```diff
@@ -75,10 +61,4 @@ public Map<String, String> handleUnknownStatusException(UnknownStatusException e
-    public Map<String, String> handleInvalidStatusException(InvalidStatusException e) {
-        return Map.of("InvalidStatusException", e.getMessage());
-    }
 }
```
Лучше еще добавить обработку необрабатываемых исключений.

sergeymamonov commented on src/main/java/ru/practicum/shareit/request/service/impl/ItemRequestServiceImpl.java:
```diff
@@ -0,0 +1,94 @@
+        List<ItemRequest> itemRequests = itemRequestRepository.findAllByRequester(requester);
+        return itemRequests.stream()
+                .map(itemRequest -> itemRequestMapper.toItemRequestDto(itemRequest,
+                        itemRepository.findAllByRequest(itemRequest.getId())
```
Запросы к БД в цикле.

sergeymamonov commented on src/main/java/ru/practicum/shareit/request/service/impl/ItemRequestServiceImpl.java:
```diff
@@ -0,0 +1,94 @@
+        List<ItemRequest> itemRequests = itemRequestRepository.findAllByRequesterNot(user, pageable);
+        return itemRequests.stream()
+                .map(itemRequest -> itemRequestMapper.toItemRequestDto(itemRequest,
+                        itemRepository.findAllByRequest(itemRequest.getId())
```
Запросы к БД в цикле.

sergeymamonov commented on src/test/java/ru/practicum/shareit/item/service/ItemIntegrationTests.java:
```diff
@@ -0,0 +1,324 @@
+
+        Long savedItemId = itemService.addItem(itemOwner.getId(), itemCreateDto).getId();
+
+        Item savedItem = itemRepository.findById(savedItemId).get();
```
Неверная практика использовать get у Optional без проверки содержимого. Хорошая статья, как работать с Optional https://struchkov.dev/blog/ru/optional-in-java/

Vexvl replied:
Спасибо за замечания, исправил всё, кроме dto

sergeymamonov reviewed (changes requested):
В качестве предложения для улучшения можно сделать объекты dto неизменяемыми классами (предложение, а не обязательное к исправлению замечание).

В тестах я бы не использовал try/catch или throws Exception в определении теста, а использовал бы аннотацию @SneakyThrows, потому что мы рассчитываем, что возникновение ошибки в тестах это редкая ситуация, и если она возникнет, то мы ее увидим. Не обязательно менять, как предложение на будущее.

Vexvl committed:
fix : BookingControllerTests.java
      BookingIntegrationTests.java
      BookingUnitTests.java
      CommentControllerTests.java
      CommentServiceImpl.java
      CommentUnitTests.java
      ExceptionsHandler.java
      ItemControllerTest.java
      ItemIntegrationTests.java
      ItemRepository.java
      ItemRequestControllerTests.java
      ItemRequestMapperTests.java
      ItemRequestServiceImpl.java
      ItemRequestTest.java
      ItemRequestUnitTest.java

style changes: RequestRepositoryTest.java
               UserControllerTests.java
               UserIntegrationTests.java
               UserMapperTests.java
               UserRepositoryTests.java
               UserUnitTests.java

sergeymamonov commented on src/main/java/ru/practicum/shareit/handler/ExceptionsHandler.java:
```diff
@@ -14,6 +15,11 @@
 public class ExceptionsHandler {
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity<String> handleUnhandledException(Exception e) {
+        return new ResponseEntity<>("An unexpected error occurred", HttpStatus.BAD_REQUEST);
```
В таком случае возвращаем 500 статус код, а не 400. 400 - это неверный запрос.

sergeymamonov commented on src/main/java/ru/practicum/shareit/request/service/impl/ItemRequestServiceImpl.java:
```diff
@@ -0,0 +1,109 @@
+        List<ItemRequest> itemRequests = itemRequestRepository.findAllByRequesterNot(user, pageable);
+
+        List<ItemRequestDto> itemRequestDtos = itemRequests.stream().map(itemRequest -> {
+                    List<Item> items = itemRepository.findAllByRequest(itemRequest.getId());
```
Запросы к БД в цикле.
